cupy/_core/_kernel.pyx

```diff
@@ -1610,6 +1636,35 @@ cdef class _Ops:
             # raise TypeError if the type combination is disallowed
             (<_Op>op).check_valid()
 
+            # check for overflow in operands. Consider `np.uint8(1) + 300`.
+            # Per NEP 50 this raises OverflowError because 300 overflows uint8.
+            # We can only check it after the output dtype is known, which is
+            # op.out_types here.
+            out_type = op.out_types[0]
+            if issubclass(out_type, numpy.integer):
+                iinfo = numpy.iinfo(out_type)
+                for i in range(len(in_args)):
+                    arg = in_args[i]
+                    typ = arg.dtype.type
+
+                    if weaks[i] and issubclass(typ, numpy.integer):
+                        if issubclass(typ, numpy.signedinteger):
+
+                            if issubclass(out_type, numpy.signedinteger):
+                                ok = (<int64_t>iinfo.min <=
+                                      <int64_t>arg <=
+                                      <int64_t>iinfo.max)
+                            else:
+                                ok = ((arg >= 0) and
+                                      (<uint64_t>arg <= <uint64_t>iinfo.max))
+
+                        elif issubclass(typ, numpy.unsignedinteger):
+                            # arg >= 0
+                            ok = <uint64_t>arg <= <uint64_t>iinfo.max
+
+                        if not ok:
+                            raise OverflowError(f"Python integer {arg} out of "
+                                                f"bounds for {out_type}.")
```

Comment:
This looks like a lot of code that really ends up doing the same as: `out_type(int(arg))` or `out_type(arg.item())`?
I suppose one could check safe-cast first, although I am not sure it would be faster.

Reply:
Not sure what you mean, mind expanding? This code is meant to check `iinfo.min <= arg <= iinfo.max` where arg can be signed or unsigned. Cython is trying to be helpful, so need explicit casts to avoid overflows.

Reply:
What I mean is, write this whole thing as:
```
for out_type, arg, weak for zip(out_types, in_args, weaks):
    if not weak is not int:  # assume we change what is stored in `weaks`
        continue

    out_type(int(in_arg))  # check if Python integer fits
```

Reply:
now that I look again, I think it also needs to loop `in_types` and not rely on `out_type` being identical.